### ISSUE TITLE
Implement qnwdist algorithm

### ISSUE THREAD

**Issue by jlperla** [QuantEcon/Expectations.jl#16]
See https://github.com/QuantEcon/QuantEcon.jl/blob/be0a32ec17d1f5b04ed8f2e52604c70c69f416b2/src/quad.jl#L918

Seems like we could add in this as a general algorithm type?  Might need to discuss with @sglyon the precise math, but the code could be merged easily

**Comment by jlperla** [comment]
The basic algorithm is to try to space grid-points so that you keep the probability mass of each cell as being the same.   This helps to make sure that you don't have a large number of grid points in areas without much probability, etc.
- The current code has two varations which you can see in the dispatch on `method` in https://github.com/QuantEcon/QuantEcon.jl/blob/be0a32ec17d1f5b04ed8f2e52604c70c69f416b2/src/quad.jl#L923
- The first method, named `Quantile` linearly spaces the nodes between the `q0` and `qN` quantile points, and then has a cell for the `[0, q0]` and the `[qN, 1]` as well.
- The second method, named `Even` is for having all of the quantile points linearly spaced.
- After getting the nodes, which is done from the `_quadnodes` function and the dispatching, it calculates the weights using the `cdf` on the distribution.
- After finding the weights and nodes, this can just return a standard `IterableExpectation` type

Why don't with the fully linear spaced one (i.e. `Even` in the current code).  Lets call the algorithm `struct QuantileLinSpace<: QuadratureAlgorithm end ` and use the `N` keyword parameter as the number of quantile points.  Otherwise, I think you can copy (and cite) the code with almost no modification.

This method can be used with any `ContinuousUnivariateDistribution` (i.e. no need to check it is bounded)
For the defaults, I would do something like
```julia
function _expectation(dist::ContinuousUnivariateDistribution, alg::Type{QuantileLinSpace}; n = 50, kwargs...)
end
```

If you have any questions, feel free to ask @sglyon

**Comment by arnavs** [comment]
@jlperla Sounds good. I will do this and post back here when done.

**Comment by arnavs** [comment]
Done. This is included and documented, and passed a few tests. Only change is `N => n`, because that's what we use in the rest of the library (although, eventually, I feel like we could let those be interchangeable, since it's such a common argument). 

Two things: 

1. @sglyon In the citations I generally say:

> Implementation of the `qnwdist()` quadrature scheme written by Spencer Lyon (PhD. NYU) as part of the [QuantEcon](http://quantecon.org) project. Used with permission. 

Let me know if you'd like me to write something else. 

2. I had some trouble running this test: 

> ` @test expectation(x -> ((x-mean_gamma)/sqrt(var_gamma))^3, gammaDist, QuantileLinSpace; n = 100, q0 = 1e-4, qN = 1-1e-4) ≈ skew_gamma atol = 1e-10`

Where `gammaDist` is the standard gamma (i.e., `Gamma()` in the Distributions package, which has mean `1.0` and variance `1.0`). Max accuracy was around 1.99 vs 2.0 with 100 nodes, varying quantiles bounds. I figure this is just the cost of using a generic quadrature scheme, but if this is behavior you're familiar with, we can put it in the docs.

**Comment by sglyon** [comment]
Great! Glad you were able to borrow the implementation. 

Citation is great, thanks for including it. 

I'm not really sure what the precise issue is with the test. Should the answer be 2.0 and the new routine gave 1.99?

**Comment by arnavs** [comment]
Right, that's the issue.

**Comment by sglyon** [comment]
Hmm. that seems decent to me, but not great.  

What does the absolute error look like as you add nodes (maybe from 10 to 100)?

**Comment by arnavs** [comment]
Here's a quick script- 

```
n = 10, err = 0.2316329373196231
n = 20, err = 0.06515754281443442
n = 30, err = 0.12898902933346923
n = 40, err = 0.14997021805364685
n = 50, err = 0.15794451214725602
n = 60, err = 0.16096792521178438
n = 70, err = 0.16186330558990503
n = 80, err = 0.16178536979788039
n = 90, err = 0.1612560800235514
n = 100, err = 0.16052737325582855
n = 110, err = 0.15972585756214452
n = 120, err = 0.15891629645937355
n = 130, err = 0.15813154889356773
n = 140, err = 0.15738750172926874
n = 150, err = 0.15669084931019173
n = 160, err = 0.1560432853943916
n = 170, err = 0.1554438194951413
n = 180, err = 0.1548900789950478
n = 190, err = 0.15437904765879362
n = 200, err = 0.15390748414972455
```

**Comment by arnavs** [comment]
Edit: It stays at about that level (above 0.14) well into n = 14,000. Maybe more. So it seems that this is less about what's happening inside the [0.001-0.999] interval, and more about what that misses.

**Comment by jlperla** [comment]
The only thing I would say is that the gamma distribution is a fat-tailed distribution, which are pretty nasty at the best of times (e.g. gauss-laguerre might be the best there).  If it works decent enough on compact-support and more reasonable infinite support ones, then maybe it just isn't a good algorithm for that PDF.

**Comment by arnavs** [comment]
Could be. And I think if you're using a distribution-agnostic quadrature algorithm, you're not exactly in a place to be complaining about accuracy (especially using default parameter values). But I figured it was worth bringing up. 

It seems, though, that this algorithm has been implemented. @jlperla and I were kicking around a slightly different implementation (set the bounds to be `0.0` and `1.0`, and then do a linspace, and then take the interior points). 

I think in terms of workflow, if it's OK with Jesse, what we can do is fix the original implementation, close this one, and then open a separate issue to test the two implementations (and to test this one generally; @sglyon, are there are unit tests for this on QuantEcon? Can't find them in the `test_quad.jl` file, ).
